### DISCUSSION TITLE
No Midi Input

### DISCUSSION THREAD

**Discussion by prof-cocoon** [marcel-licence/esp32_basic_synth#91] [Q&A]
Hi, i'm having the problem of the title: no midi activity in Serial2. I don't know if its something i'm not guessing with config.h.
Hardware seems to be ok, because uncommenting "#define NOTE_ON_AFTER_SETUP" outputs a note correctly. Also, I've made a little code to test the octocoupler and its connections to midi jack and ESP32 pin RX2 and it's monitoring ok the midi messages. Also, I've uncommented the #define MIDI_DUMP_SERIAL2_TO_SERIAL in order to see in the serial monitor any activicty, but nothing echoes. I'm using a DOIT ESP32 DEVKIT V1 board. Any help is appreciated! Thanks for your work Marcel!

This is the serial monitor output:
esp32_basic_synth  Copyright (c) 2022  Marcel Licence
This program comes with ABSOLUTELY NO WARRANTY;
This is free software, and you are welcome to redistribute it
under certain conditions; 
Initialize Synth Module
Initialize I2S Module
I2S configured using following pins:
    BCLK,BCK: 25
    WCLK,LCK: 27
    DOUT: 26
    DIN: -1
rev: 15329, 15329
Reverb is ready!
Setup Serial2 with 31250 baud with rx: 16 and tx 17
Setup MidiPort2 using Serial2
ESP.getFreeHeap() 36716
ESP.getMinFreeHeap() 36696
ESP.getHeapSize() 186856
ESP.getMaxAllocHeap() 19444
Total heap: 186856
Free heap: 36716
Total PSRAM: 0
Free PSRAM: 0
Firmware started successfully

Here is my config.h:

//#define NOTE_ON_AFTER_SETUP
#define SERIAL_BAUDRATE 115200
//#define OUTPUT_SAW_TEST

//#define BOARD_ML_V1 
//#define BOARD_ESP32_AUDIO_KIT_AC101 
//#define BOARD_ESP32_AUDIO_KIT_ES8388 
#define BOARD_ESP32_DOIT 

//#define AUDIO_PASS_THROUGH

#define SAMPLE_BUFFER_SIZE 48
#define MIDI_USE_CONST_VELOCITY
#define MAX_DELAY   (SAMPLE_RATE/2)

//#define MIDI_RECV_FROM_SERIAL
//#define MIDI_VIA_USB_ENABLED
//#define OLED_OSC_DISP_ENABLED

//#define ADC_TO_MIDI_ENABLED
//#define ADC_TO_MIDI_LOOKUP_SIZE 8 

//#define ARP_MODULE_ENABLED
//#define MIDI_SYNC_MASTER 
//#define MIDI_CTRL_ENABLED 

//#define MIDI_STREAM_PLAYER_ENABLED

**Comment by marcel-licence** [comment]
@prof-cocoon hm the output looks okay to me.
You could try to add an LED connected to VCC (+) and RX2 (-). It should blink shortly when MIDI data is transmitted.

**Reply by prof-cocoon** [reply]
Hi Marcel, thank you for your reply. I think the midi input is working ok, because i'm testing it with this little code and the serial monitor reads correctly the messages.
I'm a little lost to be honest, because I can play the midi song you provide, so the dac is ok and the sound is great. Obviously your code is right and i'm missing something...I think it's something with the serial2 configuration.

#define MIDI_SERIAL2_BAUDRATE   31250
#define RXD2 16
#define TXD2 17

void setup() {
  Serial.begin(115200);
  Serial2.begin(MIDI_SERIAL2_BAUDRATE, SERIAL_8N1, RXD2, TXD2);
}

void loop() {
    if (Serial2.available())
    {
        uint8_t incomingByte = Serial2.read();
        Serial.printf("%02x", incomingByte);
    }
}

**Comment by prof-cocoon** [comment]
Hi Marcel, good news, finally the midi messages are being received well!
This are the lines that I've modified...still I don't know why this is working and the original code don't. In the midi interface file I've commented the following lines, where the serial2 is configurated.

#ifdef MIDI_RX2_PIN
//#ifdef MIDI_TX2_PIN
//    Serial.printf("Setup Serial2 with %d baud with rx: %d and tx %d\n", MIDI_SERIAL2_BAUDRATE, MIDI_RX2_PIN, MIDI_TX2_PIN);
//    Serial2.begin(MIDI_SERIAL2_BAUDRATE, SERIAL_8N1, MIDI_RX2_PIN, MIDI_TX2_PIN);
//#else
//    Serial.printf("Setup Serial2 with %d baud with rx: %d only\n", MIDI_SERIAL2_BAUDRATE, MIDI_RX2_PIN);
//    Serial2.begin(MIDI_SERIAL2_BAUDRATE, SERIAL_8N1, MIDI_RX2_PIN);
//#endif
//    pinMode(MIDI_RX2_PIN, INPUT_PULLUP); /* can be connected to open collector output */
//#else
//    Serial.printf("Setup Serial2 with %d baud with rx: RX2 pin\n", MIDI_SERIAL2_BAUDRATE);
//    Serial2.begin(MIDI_SERIAL2_BAUDRATE);
#endif

...And in the esp32_basic_synth file I've added the following line in the setup function:
Serial2.begin(31250, SERIAL_8N1, 16, 17);

Now the fun begins...I was a little confused about the midi implementation of z_config, but now I've figured out how your edirol controller sends midi cc and I've reproduced it with a little midi out template preset in FL Studio. Everything is working well, thank you a lot for your work. It is amazing to have a polyphonic synth for such a little price, especially for our damaged economy in Argentina, where the simplest and cheapest synth costs about one month of salary.
Thank you again! a big hug! Cheers!

**Reply by marcel-licence** [reply]
Hey, 
I am happy to hear that!

Hm the behaviour of the setup is really strange. 
If you try some of the other projects you can apply the same change in case it won't work. I tried to keep the files compatible over the different projects.

You are welcome and wish you a lot of fun!

**Reply by prof-cocoon** [reply]
Hi Marcel, yes, i'm trying the same configuration in the organ project and in the drum computer project and they work flawlessly.
Also i'm adding some features to the synth project, like oscillator individual transpose, volume and pan, and filter keyboard tracking. Just trying to emulate more or less the architecture of a prophet 5. I think the most tricking trouble is the poly-mod section, i will have to study better your pwm project and fm synth.

Good sunday!

**Comment by prof-cocoon** [comment]
Good day!
I have made some mods to the original code to make the synth more oriented to soundcraft and performing live. This are the "new" features:
- Individual Oscillator Volume and Transpose
- Filter Keyboard Tracking
- LFO Rate and Waveform change (*)
- Filter Cutoff modded by LFO
- Filter Envelope Amount
- Voice Filter Cutoff knob
- Oscillator A Pitch modded by Filter Envelope

(*) Here I'm having some trouble. When the selected waveform is "continuous" (i.e sine os triangular) the mod works smooth, but when saw or square is selected, _sometimes_ there is some audible noise in the falling flank. I don't understand why this happens, as the values passed to the Filter Calculate function are legal.
Finally, I'm trying to understand your pwm example code to add that feature to this project, but I don't see where the duty cycle is being ontrolled. Another feature I want to add is to modulate the filter cutoff with the output of one oscillator, and to FM one oscillator by another, but with no success.

UPDATE: now the filter cutoff can be modulated with Osc 2 audio output, but there are still some unwanted noises. Here is a little playlist I've made with dry and wet sounds through some ValhallaDSP vsts. [Esp32 playlist](https://on.soundcloud.com/qc8ET)
